lib/tiki/torch/publishing/retries.rb

```diff
@@ -0,0 +1,82 @@
+# frozen_string_literal: true
+
+module Tiki
+  module Torch
+    module Publishing
+      class Retries
+        class Entry
+          attr_reader :topic_name, :event
+
+          def initialize(topic_name, event)
+            @topic_name = topic_name
+            @event      = event
+            @tries      = 1
+          end
+
+          def fingerprint
+            @fingerprint ||= format '%s:%s',
+                                    @topic_name.to_s,
+                                    (@event.respond_to?(:fingerprint) ? @event.fingerprint : @event.to_s)
+          end
+
+          def increment_tries
+            @tries += 1
+          end
+        end
+
+        include Logging
+
+        class << self
+          def entries
+            @entries ||= Concurrent::Hash.new
+          end
+
+          def add(topic_name, event)
+            entry = Entry.new topic_name, event
+            found = entries[entry.fingerprint]
+            if found
+              found.increment_tries
+              found
+            else
+              entries[entry.fingerprint] = entry
+            end
+          end
+
+          def executor
+            @executor ||= Concurrent::TimerTask.new(
+              execution_interval: Torch.config.retry_interval_secs,
+              timeout_interval: Torch.config.retry_timeout_secs
+            ) do
+              keys = entries.keys[0, Torch.config.retry_size]
+              keys.each do |key|
+                entry = entries.delete key
```

Comment:
you're right, need to find a better way to keep track. at the same time I'm not stopping at any retry count yet. as it is implemented it will keep retrying forever until we implement another callback to deal once the max retry is reached. it seems that it is getting overly complex though. WDYT?

Reply:
I gave it a try. although I'm not dealing with fully _dead_ jobs I'm making sure we keep them if they fail so they can be incremented. does this help?

Reply:
The documentation definitely helps.

I do like that it'll keep retrying until they go through. Under heavy load, the resource usage may grow quickly, but hopefully that situation can be avoided.

Reply:
I've added another callback to deal with an entry that has been just added. This allows to deal with too many tries on an entry or even too many entries. With this callback you could Rollbar the error or persist the message some other way.